### PR TITLE
feat(CropPage): allow skipping crop if page boundary not found (Fixes #215)

### LINKED ISSUE #215: [Enhancement] Allow skipping crop if page boundary is not visible in CropPage
**Is your feature request related to a problem? Please describe.**
In case of combined CropPage + CropOnMarkers usecase, sometimes the page boundary is not visible but the markers can still be detected. Currently in this case CropPage raises an error for the image since it thinks the image is not fit for processing further.

**Describe the solution you'd like**
We should allow a flag `shouldFailIfPageNotFound` which would be set to true by default. On setting it as false, the CropPage processor should return the input image without cropping.

**Describe alternatives you've considered**
Another alternative is to create a new processor that does this instead of a flag, but seems unnecessary as of now.

**Additional context**

- Link to the [Discord invite](https://discord.gg/qFv2Vqf)
- Link to the [Hacktoberfest 2024 channel](https://discord.com/channels/590134763784896514/1284801897714155613/1284802010549321770)

### PR DESCRIPTION
**Pull request by ChinmayyK**
Resolves #215. Added `shouldFailIfPageNotFound` to skip crop if the page is not found. Includes pre-commit formatting fixes.

### CODE REVIEW

**Review by Udayraj123 — COMMENTED** [review]
please remove unnecessary file changes and make the diff cleaner

**Comment by ChinmayyK** [comment]
> please remove unnecessary file changes and make the diff cleaner

Done, please review

File: src/processors/CropPage.py [hunk 1]
```diff
@@ -62,17 +62,27 @@ def __init__(self, *args, **kwargs):
...
+            if self.should_fail_if_page_not_found:
+                logger.error(
+                    f"\tError: Paper boundary not found for: '{file_path}'\nHave you accidentally included CropPage preprocessor?"
+                )
+                return None
+            else:
+                logger.warning(
+                    f"\tWarning: Paper boundary not found for: '{file_path}', skipping crop."
+                )
+                return original_image
```
Comment (Udayraj123) [inline src/processors/CropPage.py:85]: no need of returning image copy actually. The normalise() operation will be harmless in context of OMRChecker

**Review by Udayraj123 — COMMENTED** [review]
@ChinmayyK  please address 1 open comment, rest looks good.

File: src/processors/CropPage.py [hunk 2]
```diff
@@ -62,17 +62,27 @@ def __init__(self, *args, **kwargs):
         self.morph_kernel = tuple(
             int(x) for x in cropping_ops.get("morphKernel", [10, 10])
         )
+        self.should_fail_if_page_not_found = cropping_ops.get(
+            "shouldFailIfPageNotFound", True
+        )
 
     def apply_filter(self, image, file_path):
+        original_image = image.copy()
```
Comment (SadmaFaahiim) [inline src/processors/CropPage.py:70]: When the boundary is not found and `shouldFailIfPageNotFound` is `False`, we return `original_image` — i.e. the image *before* the `cv2.GaussianBlur` + `normalize` applied on the next line. On the success path, downstream preprocessors receive the normalized (warped) image, so a single batch can mix raw and normalized inputs depending on whether one image had a detectable boundary. That can visibly bias downstream threshold/edge/contour preprocessors and makes results file-order dependent.

Suggestion: return the already processed `image` (after the blur + normalize on line 71) instead of the raw copy — that keeps the skip path's output in the same normalization space, and lets you drop `original_image` / `image.copy()` entirely (the input isn't mutated on the success path either).

File: src/processors/CropPage.py (hunk 1 above)
Comment (SadmaFaahiim) [inline src/processors/CropPage.py:85]: Minor: the `else` is redundant after the `if ... return None` branch — early-return style (de-indent the `logger.warning` + `return`) reads cleaner and keeps the diff smaller.

File: src/tests/test_samples/sample1/boilerplate.py [hunk 3]
```diff
@@ -1,7 +1,12 @@
 TEMPLATE_BOILERPLATE = {
     "pageDimensions": [300, 400],
     "bubbleDimensions": [25, 25],
-    "preProcessors": [{"name": "CropPage", "options": {"morphKernel": [10, 10]}}],
+    "preProcessors": [
+        {
+            "name": "CropPage",
+            "options": {"morphKernel": [10, 10], "shouldFailIfPageNotFound": False},
```
Comment (SadmaFaahiim) [inline src/tests/test_samples/sample1/boilerplate.py:7]: This flips the *shared* `sample1` boilerplate to silently tolerate a missing page boundary for every sample that reuses it. That weakens the sample's regression value — a genuine boundary-detection regression could slip through CI quietly.

Rather than disabling it in the shared fixture, consider adding a dedicated unit test that covers both branches explicitly (fail path → `None`, skip path → image passes through), and keep the boilerplate at the default `True`.

File: src/schemas/template_schema.py [hunk 4]
```diff
@@ -173,7 +173,8 @@
                                     "type": "object",
                                     "additionalProperties": False,
                                     "properties": {
-                                        "morphKernel": two_positive_integers
+                                        "morphKernel": two_positive_integers,
+                                        "shouldFailIfPageNotFound": {"type": "boolean"},
```
Comment (SadmaFaahiim) [inline src/schemas/template_schema.py:177]: LGTM component: adding the option to the validated `properties` with `additionalProperties: False` means an invalid value is rejected at validation time instead of failing later. Might be worth one line in the README processor docs so users know this knob exists.

**Review by SadmaFaahiim — COMMENTED** [review]
Thanks for the change — making the boundary-not-found behavior configurable is a clean, backwards-compatible direction (the `True` default preserves existing configs). The schema is correctly gated via `additionalProperties: False`, and downgrading the log level in the skip path is appropriate.

Main concerns below are about input consistency in the skip path and test coverage for the two new branches.
